File: modules/core/src/main/scala/com/github/chengpohi/http/KibanaProxyApacheInterceptor.scala
package com.github.chengpohi.http

import org.apache.commons.lang3.StringUtils
import org.apache.http.client.methods.HttpRequestWrapper
import org.apache.http.client.utils.URIBuilder
import org.apache.http.protocol.HttpContext
import org.apache.http.{HttpRequest, HttpRequestInterceptor}

import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class KibanaProxyApacheInterceptor extends HttpRequestInterceptor {
  override def process(request: HttpRequest, context: HttpContext): Unit = {
    request match {
      case req: HttpRequestWrapper =>
        val uri = req.getURI
        val query = uri.getQuery match {
          case null => "?pretty=false"
          case f => "?" + f + (if (uri.getPath.contains("_search")) "&pretty=false" else "")
        }
        val method = Option(req.getFirstHeader("KIBANA_PROXY_METHOD")).map(i => i.getValue).getOrElse("POST");
        val pathPrefix = Option(req.getFirstHeader("KIBANA_PATH_PREFIX"))
          .map(i => i.getValue)
          .filter(i => StringUtils.isNotBlank(i) && !i.equals("/"))
          .map(i => if (!i.startsWith("/")) {
            "/" + i
          } else i)
          .getOrElse("");
        val proxyUri = new URIBuilder(uri)
          .removeQuery()
          .setPath(pathPrefix + "/api/console/proxy")
          .addParameter("path", uri.getPath.replaceAll("^/+", "") + query)
          .addParameter("method", method)
          .build();
        req.setURI(proxyUri)
        req.addHeader("kbn-xsrf", "kibana")
      case _ =>
    }
  }
}

File: src/main/java/com/github/chengpohi/http/KibanaProxyApacheInterceptor.java
package com.github.chengpohi.http;

import org.apache.commons.lang3.StringUtils;
import org.apache.http.Header;
import org.apache.http.HttpRequest;
import org.apache.http.HttpRequestInterceptor;
import org.apache.http.client.methods.HttpRequestWrapper;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.protocol.HttpContext;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.regex.Pattern;

public class KibanaProxyApacheInterceptor implements HttpRequestInterceptor {

    private static final Pattern LEADING_SLASHES_PATTERN = Pattern.compile("^/+");

    @Override
    public void process(HttpRequest request, HttpContext context) {
        if (request instanceof HttpRequestWrapper) {
            HttpRequestWrapper wrapper = (HttpRequestWrapper)request;
            URI uri = wrapper.getURI();

            String query = uri.getQuery();
            String queryString;
            if (query != null) {
                queryString = "?" + query + (uri.getPath().contains("_search") ? "&pretty=false" : "");
            } else {
                queryString = "?pretty=false";
            }

            String method = getHeaderValue(wrapper, "KIBANA_PROXY_METHOD");
            if (method == null) {
                method = "POST";
            }

            String pathPrefix = getHeaderValue(wrapper, "KIBANA_PATH_PREFIX");
            String finalPathPrefix = "";
            if (StringUtils.isNotBlank(pathPrefix) && !"/".equals(pathPrefix)) {
                finalPathPrefix = pathPrefix.startsWith("/") ? pathPrefix : "/" + pathPrefix;
            }

            URIBuilder proxyUriBuilder = new URIBuilder(uri);
            proxyUriBuilder.removeQuery();
            proxyUriBuilder.setPath(finalPathPrefix + "/api/console/proxy");
            proxyUriBuilder.addParameter("path", LEADING_SLASHES_PATTERN.matcher(uri.getPath()).replaceAll("") + queryString);
            proxyUriBuilder.addParameter("method", method);

            try {
                wrapper.setURI(proxyUriBuilder.build());
                wrapper.addHeader("kbn-xsrf", "kibana");
                wrapper.addHeader("osd-xsrf", "true");
            } catch (URISyntaxException e) {
                throw new RuntimeException("Failed to build proxy URI", e);
            }
        }
    }

    private String getHeaderValue(HttpRequestWrapper request, String headerName) {
        Header header = request.getFirstHeader(headerName);
        return header != null ? header.getValue() : null;
    }
} 